ttnn/cpp/ttnn/operations/data_movement/common/kernels/common.hpp

```diff
@@ -6,9 +6,82 @@
 // It's best to copy and paste the functions in rather than include the header as code size will likely explode
 // Best to separate in to cpp/hpp at some point to avoid the code size explosion but need to figure out the linking
 // issues
+#include <stdio.h>
+#include <cstring>
+#define MASK_64      0xFFFFFFFFFFFFFFC0
+#define OFFSET_64    0x000000000000003F
+#define MASK_16      0xFFFFFFFFFFFFFFF0
+#define OFFSET_16    0x000000000000000F
 
 namespace tt::data_movement::common {
 
+#define max_packet_size 8192
+
+template <uint32_t max_transfer_size, bool only_reads>
+FORCE_INLINE void enhanced_noc_async_read(
+    const uint64_t src_noc_addr, const uint32_t dst_l1_addr, const uint32_t bytes) {
+    // If you do not know the max_transfer_size at compile time write 0 to it.
+    // only reads is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as True
+    if constexpr (((max_transfer_size < max_packet_size) && (max_transfer_size != 0)) || only_reads) {
+        noc_async_read_one_packet(src_noc_addr, dst_l1_addr, bytes);
+    } else {
+        noc_async_read(src_noc_addr, dst_l1_addr, bytes);
+    }
+}
+
+template <uint32_t max_transfer_size, bool only_writes>
+FORCE_INLINE void enhanced_noc_async_write(
+    const uint32_t src_l1_addr, const uint64_t dst_noc_addr, const uint32_t bytes) {
+    // If you do not know the max_transfer_size at compile time write 0 to it.
+    // only writes is true if we ONLY use noc_async_read and all calls to tt_memmove have use_read_datamover as False
+    if constexpr (((max_transfer_size < max_packet_size) && (max_transfer_size != 0)) || only_writes) {
+        noc_async_write_one_packet(src_l1_addr, dst_noc_addr, bytes);
+    } else {
+        noc_async_write(src_l1_addr, dst_noc_addr, bytes);
+    }
+}
+
+template <bool guaranteed_16B_alligned, bool copy_async, bool use_read_datamover, uint32_t max_transfer_size>
```

Comment:
A number of typos alligned

Reply:
Lol, English not being my first language is showing